File: artifacts/libzahl-1.0/decompose/src/zand.c
void zand(z_t a, z_t b, z_t c)
{
  size_t n;
  if (zzero(b) || zzero(c))
  {
    a->sign = 0;
    return;
  }
  n = (b->used < c->used) ? (b->used) : (c->used);
  while (n--)
    if (b->chars[n] & c->chars[n])
  {
    goto found_highest;
  }

  a->sign = 0;
  return;
  n += 1;
  found_highest:
  a->used = n;

  if (a == b)
  {
    while (n--)
      a->chars[n] &= c->chars[n];

  }
  else
    if (a == c)
  {
    while (n--)
      a->chars[n] &= b->chars[n];

  }
  else
  {
    do
    {
      if (a->alloced < a->used)
      {
        libzahl_realloc(a, a->used);
      }
    }
    while (0);
    memcpy(a->chars, c->chars, a->used * (sizeof(zahl_char_t)));
    while (n--)
      a->chars[n] &= b->chars[n];

  }
  a->sign = (((zsignum(b) > 0) || (zsignum(c) > 0)) * 2) - 1;
}


/*** DEPENDENCIES:
inline static int zzero(z_t a)
{
  return !a->sign;
}


----------------------------
typedef struct 
{
  int sign;
  size_t used;
  size_t alloced;
  zahl_char_t *chars;
} z_t[1]
----------------------------
inline static int zsignum(z_t a)
{
  return a->sign;
}


----------------------------
void libzahl_realloc(z_t a, size_t need)
{
  size_t i;
  size_t x;
  zahl_char_t *new;
  unsigned int new_idx = 0;
  if ((need & ((~need) + 1)) != need)
  {
    need |= need >> 1;
    need |= need >> 2;
    need |= need >> 4;
    for (i = sizeof(need), x = 8; i; i >>= 1, x <<= 1)
      need |= need >> x;

    need += 1;
  }
  for (i = 0, x = need; x; x >>= 1)
    i += 1;

  if (libzahl_pool_n[i])
  {
    libzahl_pool_n[i] -= 1;
    new_idx = libzahl_pool[i][libzahl_pool_n[i]];
    memcpy(new, a->chars, a->alloced * (sizeof(zahl_char_t)));
    zfree(a);
    a->chars = &new[new_idx];
  }
  else
  {
    a->chars = realloc(a->chars, need * (sizeof(zahl_char_t)));
    if (!a->chars)
    {
      if (!errno)
      {
        errno = ENOMEM;
      }
      libzahl_error = errno, longjmp(libzahl_jmp_buf, 1);
    }
  }
  a->alloced = need;
}


----------------------------
***/

File: artifacts/libzahl-1.0/translate/src/zand.rs
use crate::*;

pub fn zand(a: &mut Z, b: &Z, c: &Z) {
    if zzero(b) || zzero(c) {
        a.sign = 0;
        return;
    }

    let n = if b.used < c.used { b.used } else { c.used };
    let mut found_n = Option::None;

    for i in (0..n).rev() {
        if let (Some(b_chars), Some(c_chars)) = (&b.chars, &c.chars) {
            if (b_chars[i] & c_chars[i]) != 0 {
                found_n = Some(i + 1);
                break;
            }
        }
    }

    let n = match found_n {
        Some(n) => n,
        None => {
            a.sign = 0;
            return;
        }
    };

    a.used = n;

    if std::ptr::eq(a, b) {
        if let (Some(a_chars), Some(c_chars)) = (&mut a.chars, &c.chars) {
            for i in 0..n {
                a_chars[i] &= c_chars[i];
            }
        }
    } else if std::ptr::eq(a, c) {
        if let (Some(a_chars), Some(b_chars)) = (&mut a.chars, &b.chars) {
            for i in 0..n {
                a_chars[i] &= b_chars[i];
            }
        }
    } else {
        if a.alloced < a.used {
            libzahl_realloc(a, a.used);
        }

        if let (Some(a_chars), Some(b_chars), Some(c_chars)) = (&mut a.chars, &b.chars, &c.chars) {
            a_chars[..n].copy_from_slice(&c_chars[..n]);
            for i in 0..n {
                a_chars[i] &= b_chars[i];
            }
        }
    }

    a.sign = (((zsignum(b) > 0) || (zsignum(c) > 0)) as i32) * 2 - 1;
}
